In [1]:
# preamble
from datetime import datetime
start = datetime.now()

## Step 3 > In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved.

In [2]:
#import libraries 
import pandas as pd
import numpy as np
import os

#define jupyter notebook's root path 
ICpath = r'C:\Users\joseh\Instacart Basket Analysis'

In [3]:
#import the pickle file prepared in the PART 1
df_ords_prods_combined = pd.read_pickle(os.path.join(ICpath,'02 Data','Prepared Data','order_products_combined.pkl'))

In [4]:
#verifying columns and size
df_ords_prods_combined

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                  0   2539329        1             1          2   
1                  0   2539329        1             1          2   
2                  0   2539329        1             1          2   
3                  0   2539329        1             1          2   
4                  0   2539329        1             1          2   
...              ...       ...      ...           ...        ...   
32434484     3421081   2977660   206209            13          1   
32434485     3421081   2977660   206209            13          1   
32434486     3421081   2977660   206209            13          1   
32434487     3421081   2977660   206209            13          1   
32434488     3421081   2977660   206209            13          1   

          order_hour_of_day  days_since_prior_order  \
0                         8                     NaN   
1                         8                     NaN   
2                         8                     NaN   
3                         8                     NaN   
4                         8                     NaN   
...                     ...                     ...   
32434484                 12                     7.0   
32434485                 12                     7.0   
32434486                 12                     7.0   
32434487                 12                     7.0   
32434488                 12                     7.0   

          first_order_from_customer  product_id  add_to_cart_order  reordered  \
0                              True         196                  1          0   
1                              True       14084                  2          0   
2                              True       12427                  3          0   
3                              True       26088                  4          0   
4                              True       26405                  5          0   
...                             ...         ...                ...        ...   
32434484                      False       14197                  5          1   
32434485                      False       38730                  6          0   
32434486                      False       31477                  7          0   
32434487                      False        6567                  8          0   
32434488                      False       22920                  9          0   

         _merge  
0          both  
1          both  
2          both  
3          both  
4          both  
...         ...  
32434484   both  
32434485   both  
32434486   both  
32434487   both  
32434488   both  

[32434489 rows x 12 columns]

## Step 4 > Check the shape of the imported dataframe (it should be the same as the one you exported—always check!).

In [5]:
df_ords_prods_combined.shape

(32434489, 12)

verified the shape of the imported dataframe is 32,434,489 rows by 12 columns

## Step 5 > Determine a suitable way to combine the ```orders_products_combined``` dataframe with your products data set. Make sure you’re using your wrangled, cleaned, and deduped products data set stored in your “Prepared Data” folder from the previous Exercise’s task.

In [6]:
#first it is necessary to import the wrangled/clean version of the products dataset
df_prods = pd.read_csv(os.path.join(ICpath, '02 Data', 'Prepared Data', 'products_checked.csv'))
df_prods.dtypes

Unnamed: 0         int64
product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

In [7]:
#eliminating an unnecessary column which used to be an index before the cleanning/wranglign took place
df_prods2 = df_prods.drop(columns=['Unnamed: 0'])
df_prods2.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

In [8]:
#verifying columns and shape
df_prods2

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49667       49684          Vodka, Triple Distilled, Twist of Vanilla   
49668       49685                 En Croute Roast Hazelnut Cranberry   
49669       49686                                   Artisan Baguette   
49670       49687         Smartblend Healthy Metabolism Dry Cat Food   
49671       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49667       124              5     5.3  
49668        42              1     3.1  
49669       112              3     7.8  
49670        41              8     4.7  
49671        73             11    13.5  

[49672 rows x 5 columns]

### From steps 3 and 5 we can see it is possible to merge (join) the datasets based on ```product_id```

In [ ]:
# prepare df_ords_prods_combined for the merge (join) operation by
# first, dropping the column '_merge' showing the results from a previous merge operation
# second, dropping the column 'Unnamed: 0' which was an index column before a prior merge operation
df_ords_prods_combined_2 = df_ords_prods_combined.drop(columns=['Unnamed: 0','_merge'])

#merge the two data sets on product_id
df_ords_prods_details_combined = df_ords_prods_combined_2.merge(df_prods2, on = 'product_id', indicator = True)

In [ ]:
df_ords_prods_combined_2.shape

## Step 6 > Confirm the results of the merge using the merge flag

In [ ]:
#confirming the results based on the column '_merge'
df_ords_prods_details_combined['_merge'].value_counts()

In [ ]:
#checking the shape of the new dataframe
df_ords_prods_details_combined.shape

In [ ]:
#getting a quick view of top/bottom entries and size
df_ords_prods_details_combined

## Step 7 > Export the newly created dataframe as ```orders_products_merged``` in a suitable format (taking into consideration the size)

In [ ]:
#the best export format is pickle given the size of the file

#exporting the dataframe in pickle format
df_ords_prods_details_combined.to_pickle(os.path.join(ICpath,
                                                      '02 Data',
                                                      'Prepared Data',
                                                      'orders_products_merged.pkl'
                                                     )
                                        )

In [ ]:
# fyi only - to track performance of this notebook 
print(' time to run this notebook h:mm:ss: ', datetime.now() - start)

In [ ]:
df_ords_prods_details_combined.shape

In [ ]:
df_ords_prods_details_combined.head(3)